### PR TITLE
[profiler] Rework GC roots reporting.

### PR DESCRIPTION
**Pull request by luhenry**
This change re-implement GC roots reporting in a way that allows users to correctly track the source of a GC root.

The new design is based on emitting GC root information upfront and only report addresses during heap dumps.

When a GC root is registered we emit an event that is an address range, the root kind, a key and a text description.

A decoder can use this information to match a reported root address with all registration addresses to figure out what they mean.

A GC Root key is used to further its meaning. For example, if kind is thread, key is a tid, if kind is static variables, key is the class pointer and so on.

The build of the change is introducing this key argument across all root registration code and rework our root reporting code to encode their addresses.

Some roots use pseudo-addresses when they are not registrable. This is the case of thread stacks/registers and the finalizer queue.

Finally, root reporting was changed to happen only once per collection and at the end, leading to correct and useful data being produced.

### CODE REVIEW

**Comment by luhenry** [comment]
This is https://github.com/mono/mono/pull/5126 with master merged in to resolve conflicts.

**Comment by alexrp** [comment]
build profiler stress

**Comment by alexrp** [comment]
build profiler stress

**Comment by luhenry** [comment]
build profiler stress

File: mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogProcessor.cs [hunk 1]
```diff
@@ -414,17 +414,13 @@ LogEvent ReadEvent ()
...
 
-					hre.MaxGenerationCollectionCount = (long) _reader.ReadULeb128 ();
-
 					for (var i = 0; i < list.Length; i++) {
 						list [i] = new HeapRootsEvent.HeapRoot {
-							ObjectPointer = ReadObject (),
-							Attributes = StreamHeader.FormatVersion == 13 ? (LogHeapRootAttributes) _reader.ReadByte () : (LogHeapRootAttributes) _reader.ReadULeb128 (),
-							ExtraInfo = (long) _reader.ReadULeb128 (),
+							AddressPointer = ReadPointer (),
+							ObjectPointer = ReadObject ()
```
Comment (rodrmoya) [inline mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogProcessor.cs]: This is breaking support for v13, isn't it?

**Comment by DavidKarlas** [comment]
@monojenkins build pkg
I want to do some testing locally profiling VSfM, local build of mono doesn't have GTK and MSBuild assemblies... hence I need .pkg

**Comment by DavidKarlas** [comment]
@monojenkins build pkg

**Comment by luhenry** [comment]
build

File: mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogProcessor.cs (hunk 1 above)
Comment (luhenry) [reply mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogProcessor.cs]: This is something I missed from merging master, I will fix that.

**Comment by luhenry** [comment]
build profiler stress

**Comment by luhenry** [comment]
build pkg

**Comment by DavidKarlas** [comment]
Jetlag doesn't let me sleep...
So with help from @alexrp I finally got my custom profiler working to be loaded at startup...
I will explain how my profiler work:
- Start profiling monodevelop via `LD_LIBRARY_PATH=/Users/davidkarlas/GIT/Krofiler/Krofiler.PrototypingGround/bin/Debug/ mono --profile=krofiler main/build/bin/MonoDevelop.exe --no-redirect` LD_LIBRARY_PATH to find my krofiler lib...
- This starts logging all new register/unregister events from this PR: https://gist.github.com/DavidKarlas/a27ff3b4deddd1f23ef0812ad9fe18fd#file-gistfile1-txt-L412 https://gist.github.com/DavidKarlas/a27ff3b4deddd1f23ef0812ad9fe18fd#file-gistfile1-txt-L384-L385
- Later when whole IDE is started I call `krofiler_take_heapshot`(via PInvoke from C#) https://gist.github.com/DavidKarlas/a27ff3b4deddd1f23ef0812ad9fe18fd#file-gistfile1-txt-L391
- What this command is supposed to do... Register to capture all roots visited during GC collection via `mono_profiler_set_gc_roots_callback` and when `MONO_GC_EVENT_PRE_START_WORLD` event happens do `mono_gc_walk_heap`

Issue that I have...
Via `mono_profiler_set_gc_roots_callback` I receive 224,902 **unique** roots
Via new(from this PR) `mono_profiler_set_gc_root_register_callback` I get 1886 roots registrations
Via new(from this PR) `mono_profiler_set_gc_root_unregister_callback` I get 257 root unregistrations
And I also get 788,591 objects via heap_walk.

So my issue is... Which number is off here? New API registering too little stuff or `mono_profiler_set_gc_roots_callback` too much? Or am I just using it wrong?

**Comment by DavidKarlas** [comment]
I'm trying to make sense of all this data without much success...

Would be nice if some unit tests could be added for example... profile app which creates object and assign it to static variable... In unit test verify that root is reported...

**Comment by luhenry** [comment]
build profiler stress

**Comment by luhenry** [comment]
@kumpera we still have the profiler stress test crash at https://jenkins.mono-project.com/job/test-mono-pull-request-profilerstress/20/testReport/MonoTests/profiler-stress/msbiology/

**Comment by kumpera** [comment]
Ugg, the stress test actually caught a bug. Relevant dump: ```
Thread 2 (Thread 0x2afae280b700 (LWP 29398)):
#0  0x00002afac7892ed9 in waitpid () from /lib/x86_64-linux-gnu/libpthread.so.0
#1  0x00000000004b513e in mono_handle_native_crash (signal=0x6c2fe6 "SIGABRT", ctx=<optimized out>, info=<optimized out>) at mini-exceptions.c:2726
#2  <signal handler called>
#3  0x00002afac7cedc37 in raise () from /lib/x86_64-linux-gnu/libc.so.6
#4  0x00002afac7cf1028 in abort () from /lib/x86_64-linux-gnu/libc.so.6
#5  0x0000000000684f19 in mono_log_write_logfile (log_domain=<optimized out>, level=<optimized out>, hdr=<optimized out>, message=0x2afaf8005310 "init_thread: failed to insert thread 0x2afae280b700 in log_profiler.profiler_thread_list, found = true") at mono-log-common.c:135
#6  0x00000000006994f0 in monoeg_g_logv (log_domain=0x0, log_level=G_LOG_LEVEL_ERROR, format=<optimized out>, args=args@entry=0x2afae280aa08) at goutput.c:115
#7  0x00000000006995a2 in monoeg_g_log (log_domain=log_domain@entry=0x0, log_level=log_level@entry=G_LOG_LEVEL_ERROR, format=format@entry=0x2afac80a3838 "%s: failed to insert thread %p in log_profiler.profiler_thread_list, found = %s") at goutput.c:125
#8  0x00002afac808c77d in init_thread (add_to_lls=add_to_lls@entry=1) at log.c:547
#9  0x00002afac808c7ba in init_thread (add_to_lls=<optimized out>) at log.c:567
#10 get_thread () at log.c:569
#11 0x00002afac808e69d in get_thread () at log.c:1524
#12 gc_handle (op=0, type=MONO_GC_HANDLE_NORMAL, handle=107, obj=0x2afac709a510, prof=<optimized out>) at log.c:1524
#13 0x00000000005d864b in mono_profiler_raise_gc_handle_created (handle=handle@entry=107, type=MONO_GC_HANDLE_NORMAL, object=object@entry=0x2afac709a510) at ../../mono/metadata/profiler-events.h:72
#14 0x000000000062c616 in sgen_client_gchandle_created (handle_type=<optimized out>, obj=obj@entry=0x2afac709a510, handle=handle@entry=107) at sgen-mono.c:2964
#15 0x0000000000650b00 in alloc_handle (handles=handles@entry=0x9cc370 <gc_handles+624>, obj=obj@entry=0x2afac709a510, track=0) at sgen-gchandles.c:190
#16 0x0000000000650e41 in sgen_gchandle_new (obj=obj@entry=0x2afac709a510, pinned=pinned@entry=0) at sgen-gchandles.c:235
#17 0x000000000062c275 in mono_gchandle_new (obj=obj@entry=0x2afac709a510, pinned=pinned@entry=0) at sgen-mono.c:2841
#18 0x00000000005e4e89 in mono_thread_attach_internal (thread=0x2afac840b400, force_attach=0, force_domain=force_domain@entry=0) at threads.c:673
#19 0x00000000005e7e70 in start_wrapper_internal (stack_ptr=0x2afae280c000, start_info=0x26775c0) at threads.c:944
#20 start_wrapper (data=0x26775c0) at threads.c:1069
#21 0x00002afac788b184 in start_thread () from /lib/x86_64-linux-gnu/libpthread.so.0
#22 0x00002afac7db4ffd in clone () from /lib/x86_64-linux-gnu/libc.so.6```

**Comment by luhenry** [comment]
@kumpera could you please work on getting the profiler stress test fixed? I will rebase on master, fix the conflicts and push back here. Feel free to push anything you want to `luhenry/mono:gc_roots_experiments`

**Comment by luhenry** [comment]
build profiler stress

**Comment by DavidKarlas** [comment]
@monojenkins build pkg

**Comment by DavidKarlas** [comment]
@monojenkins build pkg

**Comment by luhenry** [comment]
@monojenkins build profiler stress

**Comment by DavidKarlas** [comment]
@monojenkins build pkg

**Comment by DavidKarlas** [comment]
https://github.com/mono/mono/blob/master/mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogStreamHeader.cs#L12 needs to be bumped to 15

**Comment by DavidKarlas** [comment]
This can be closed since it was reopened and merged as another PR.

**Pull request closed without merging**